shared/vendored/snakemake/versioning.smk

```diff
@@ -0,0 +1,191 @@
+        This uses importlib.metadata to check the available version which avoids importing
+        the top-level import.
+
+        If the package is found but the version doesn't satisfy the provided *specifier*
+        we log an error. Raises `PackageNotFoundError` if the package is not found.
+        """
+        try:
+            if name=='snakemake':
+                # in conda environments importlib reports a snakemake version of 0.0.0,
+                # so follow the approach of Snakemake's own min_version function
+                version = Version(snakemake_version)
+            else:
+                version = Version(importlib_version(name))
+        except InvalidVersion: # <https://packaging.pypa.io/en/stable/version.html#packaging.version.InvalidVersion>
+            self.unexpected_errors.append(f"Python dependency {name!r} reported a version of {output} which we were unable to parse")
+            return
+
+        ok = specifier.contains(version)
+        # print(f"[DEBUG] Checking python dependency: {name!r} installed: {version} requirements: {specifier} OK? {ok}", file=stderr)
+        if not ok:
+            self.version_incompatibilities.append(f"Python dependency {name!r} version incompatibility. You have {version} but this pathogen declares {specifier}")
+
+    def check_cli_version(self, name: str, specifier: SpecifierSet) -> None:
+        """
+        Check whether the requested *name* is (a) installed and (b) reports a version
+        which satisfies the *specifier*. Both (a) and (b) are achieved by calling
+        `<name> --version`.
+        
+        If *name* isn't found (or is not executable) we raise a ProgramNotFoundError.
+        If the package is found but the version doesn't satisfy the provided *specifier*
+        we log an error. 
+        """
+        if which(name) is None:
+            raise ProgramNotFoundError()
+
+        cmd = [name, "--version"]
+        try:
+            proc = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, check=True)
+            output = ((proc.stdout or "") + " " + (proc.stderr or "")).strip()
+        except subprocess.CalledProcessError as e:
+            self.unexpected_errors.append(f"CLI program {name!r} exited code {e.returncode} when called using {' '.join(cmd)!r}")
+            return
+
+        m = re.search(r"\d+(\.\d+(\.\d+)?)?([.-][0-9A-Za-z]+)*", output)
+        #                1   . 2   . 3           alpha etc
+        if not m:
+            self.unexpected_errors.append(f"CLI program {name!r} didn't report a parseable version when called via {' '.join(cmd)!r}")
+            return
+
+        try:
+            version = Version(m.group(0))
+        except InvalidVersion: # <https://packaging.pypa.io/en/stable/version.html#packaging.version.InvalidVersion>
+            self.unexpected_errors.append(f"CLI program {name!r} reported a version of {m.group(0)} which we were unable to parse")
```

Comment:
This will fail for versions which don't conform to Python's idea of a version, of course. Non-Python CLI tools seem more likely to hit that. FWIW, Nextstrain CLI has code for parsing any version string in a way that leaves the version at least comparable to like versions.